internal/ethapi/celo_block.go

```diff
@@ -0,0 +1,154 @@
+package ethapi
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+var (
+	gasPriceMinimumABIJson = `[{"inputs":[{"internalType":"uint256","name":"gasPriceMinimum","type":"uint256"}],"name":"GasPriceMinimumUpdated","outputs":[],"type":"event"}]`
+	gasPriceMinimumABI     abi.ABI
+)
+
+func init() {
+	parsedAbi, _ := abi.JSON(strings.NewReader(gasPriceMinimumABIJson))
+	gasPriceMinimumABI = parsedAbi
+}
+
+// PopulatePreGingerbreadBlockFields populates the baseFee and gasLimit fields of the block for pre-gingerbread blocks
+func PopulatePreGingerbreadBlockFields(ctx context.Context, backend CeloBackend, block *types.Block) *types.Block {
+	newHeader := PopulatePreGingerbreadHeaderFields(ctx, backend, block.Header())
+	block = block.WithSeal(newHeader)
+	return block
+}
+
+// PopulatePreGingerbreadHeaderFields populates the baseFee and gasLimit fields of the header for pre-gingerbread blocks
+func PopulatePreGingerbreadHeaderFields(ctx context.Context, backend CeloBackend, header *types.Header) *types.Header {
+	// If the block is post-Gingerbread, return the header as is
+	if backend.ChainConfig().IsGingerbread(header.Number) {
+		return header
+	}
+
+	var (
+		gasLimit *big.Int
+		baseFee  *big.Int
+	)
+
+	if chainId := backend.ChainConfig().ChainID; chainId != nil {
+		gasLimit = retrievePreGingerbreadGasLimit(chainId.Uint64(), header.Number)
+	}
+
+	baseFee, err := rawdb.ReadPreGingerbreadBlockBaseFee(backend.ChainDb(), header.Hash())
+	if baseFee == nil {
+		if err != nil {
+			log.Debug("failed to load pre-Gingerbread block base fee from database", "block", header.Number.Uint64(), "err", err)
+		}
```

Comment:
I would handle the error before checking for a nil base fee.
```suggestion
	baseFee, err := rawdb.ReadPreGingerbreadBlockBaseFee(backend.ChainDb(), header.Hash())
	if err != nil {
		log.Debug("failed to load pre-Gingerbread block base fee from database", "block", header.Number.Uint64(), "err", err)
	}
	if baseFee == nil {
```

Reply:
Fixed at dac08dbd886c9674fc6fc55499a45986ff007274